V10/cmd/post/man/post.1
'\" t
.TH POST 1 "Post 4.0" " "
.ds q \fBpost\fP
.ds p \fIpost\fP
.SH NAME
post \- read or send mail with corporate directory access
.SH SYNOPSIS
.SS Reading Mail
\*q [\fImailx\fR(1) options]
.SS Sending Mail
\*q [\fImailx\fR(1) options] [directory query options] address ...
.SS Directory Assistance
\*q \fB\-D\fR
.br
\*q \fB\-w\fR [directory query options] address ...
.SS Directory Update
\*q \fB\-S\fR
.SS Administrative
\*q \fB\-G\fR [\fIsource\fR [\fIdest\fR]]
.br
\*q \fB\-V\fR
.SH DESCRIPTION
\*p is a mail processing system for reading and sending
messages using
.IR mailx (1).
It provides the ability to look up employee/site information in 
on-line corporate directories, in addition
to recognizing
.IR mailx (1)
and
.IR mail (1)
addressing conventions.
\*p also generates paper mail for intra-company delivery.
.SS "Addressing"
An
.I address
is one of:
.RS 2
.TS
tab(;);
l l.
\fIalias\fP;as defined in \fImailx\fR(1)
\fIlocal file delivery\fP;\fB+\fIfile\fR or \fB|\fIprogram\fR as defined in \fImailx\fR(1)
\fInetwork address\fP;[\fIsystem\fB!\fR...]\fB!\fIlogin\fR or \fIlogin\fB@\fIsystem\fR
\fIpaper address\fP;\fBpaper!\fIname\fB:\fIloc\fB:\fIroom\fR
\fIlogin\fP;local user login id
\fIquery\fP;corporate directory query
.TE
.RE
.PP
First, 
.I alias
is resolved.
If the first character is a \fB+\fR or \fB|\fR, \fIlocal file delivery\fP
is assumed.
Then, \fIaddress\fP is searched from the left for the
first occurrence of a \fB!\fR, \fB@\fR, \fB/\fR, \fB:\fP, or \fB\=\fR.
If this character is a
.BR ! " or"
.BR @ ,
.I address
is a 
.I "network address"
.RI ( login
in
.I "network address"
can be a
.I query
or
.IR mail (1)
system-wide alias).
An
.I address
beginning with
.B paper!
delivers intra company mail specified by
.IR name ,
.IR loc ,
.RI and " room.
If the first character from the left is a
.BR / ,
.BR : " or"
.BR = ,
then
.I address
is a
.I query.
Finally, if
.B \%POSTETC
is set in the environment and
.I address
is found in
.BR /etc/passwd ,
it is a
.IR login ,
otherwise it is treated as a
.IR query .
A
.I query
is one more
.I "value pairs"
separated by
.BR / " or"
.BR : .
Colon must be used in place of slash in \fIsystem\fB!\fIquery\fR.
A
.I "value pair" 
is the construction \fIattribute\fB=\fIvalue\fR
(see
.BR Attributes ).
.SS "Reading Mail"
.PP
See
.IR mailx (1).
.SS "Sending Mail"
See
.IR mailx (1)
and
.B Directory Queries.
A resolution prompt will appear if
.I address
is ambiguous or invalid.
.br
.SS "Directory Assistance"
.TP
.B \-D
Gives you information on the corporate directory:
a support contact for reporting problems,
the last update of the directory,
and available attributes and titles.
.TP
.B \-w
Who.
Query directory.
.br
.ne 2i
.SS "Directory Update"
.TP
.B \-S
Update user information interactively and set up a new user on the system.
This is automatically invoked when using \*p for the first time to send mail.
The session prompts for personnel identification number,
reporting organization,
location code, room number, telephone number and electronic
mail address.
The default for each prompt is displayed within brackets.
To delete a value, enter
.BR \- .
Any new information is electronically mailed to the Corporate Directory
Administrator for update of the master directory.
.SS "Administrative"
.TP
.B \-G \fR[\fIsource\fR [\fIdest\fR]]
Generate a new
.IR user (4)
file by reading the
.I source
file, deleting entries for users no longer on the system, and writing
the result to the file
.IR dest .
Both
.I source
and 
.I dest
are in
.IR user (4)
format and the default for both is
.BR $POST/lib/user .
.TP
.B \-V
Print the current version and release date.
.SS "Directory Queries"
.TP
.BI \-a " attribute\fR[\fI/attribute\fR...]
Provide default
.IR attribute(s) .
If a
.I query
sequence is missing an
.IR attribute ,
assume the corresponding
.I attribute
in the
.B \-a
argument.
The default is
.BR pn .
.TP
.BI \-m " mods
Opens the directory specified by
.I mods
(see
.IR modules (4)).
.TP
.BI \-o " fmt
Format output of directory search (see
.BR "Output Formatting" ).
The default format is
.br
.B "%24pn %10org %6loc %6room %12tel %ema"
.TP
.BI \-q " query
Provide default value pairs.
If a
.I query
is missing a
value pair given in the
.BR \-q
argument, add the value pair to the
.I query
(see
.BR POSTQUAL ).
.br
.ne 8i
.SS Attributes
The ``Directory Attributes'' table below contains a list of the
attributes used in
.IR queries .
The table also supplies a short description, a match type and a list
of value characteristics (see explanation following table) for each attribute.
.PP
.TS
center,allbox,tab(;); 
c s s s s
c | c | c | c | c
l | l | c | c | l.
\fBDirectory Attributes\fP
\fIAttribute;Description;Match Type;Characteristics;Example\fR
_
pn;personal name;x;i;pn=j.f.s.smith,jr
name;personal name;x;i;name=p_jones_\^_md
first;first name;p;ia;first=frederick
middle;first middle name;p;ia;middle=bernard
middle2;second middle name;p;ia;middle2=emile
last;last name;s;ia;last=adams
suffix;name suffix;e;ia;suffix=sr
soundex;soundex code;e;i;soundex=m460
_
pid;personnel identification number;n;a;pid=103459
org;organization code;p;i;org=45263
com;company affiliation;p;i;com=bl
tl;title abbreviation;w;i;tl=dh
occ;occupational code;e;;occ=c
_
tel;full telephone number;t;a;tel="(201) 386-4419"
area;area code;e;a;area=201
exch;exchange;e;a;exch=386
ext;extension;e;a;ext=4419
cornet;cornet number;e;a;cornet=232
_
loc;location code;s;ia;loc=wh
room;room number;p;ia;room=3a-359
street;street address;p;ia;street="whippany road"
city;city address;p;ia;city=whippany
state;state address;e;ia;state=nj
zip;zip code;p;a;zip=07981
_
ema;email address;s;;ema=attmail!fsmith
_
multi;employees with multiple locations;e;m;multi=1
_
attribute;display valid attributes;e;;\-o %attribute
.TE
.PP
.SS Match Type
.PD 0
.TP
.B e
Exact match.
.TP
.B n 
Numeric match.
.TP
.B p
Prefix match.
.TP
.B s
Exact match unless argument ends with a star
.RB ( * ),
in which case prefix match.
.TP
.B t
Telephone match.
Treat rightmost four numbers as extension,
next three rightmost as exchange,
and last three rightmost as area code.
Non-alphanumerics are ignored.
The output format is
.IR "area exch ext" " ."
.TP
.B w
Prefix match on every word.
.TP
.B x
Personal name match.
Syntax for personal name is:
[\fIfirst\fB.\fR[\fImiddle\fB.\fR[\fImiddle2\fB.\fR]]]\fIlast\fR[\fB,\^\fIsuffix\fR]\ \ or
[\fIfirst\fB_\fR[\fImiddle\fB_\fR[\fImiddle2\fB_\fR]]]\fIlast\fR[\fB_\^_\fIsuffix\fR]
.br
.ne 1i
To prefix match on last name, append
.B *
or
.B ...
to
.IR last .
To phonetic match on last name, append
.B ?
or
.B ~
to
.IR last ,
or prepend personal name with
.BR ? .
The output format for
.B pn
is
\fIlast\fB,\ \fIfirst\ middle\ middle2\fB,\ \fIsuffix\fR\ ;
and
.B name
is
\fIfirst\fB_\fImiddle\fB_\fImiddle2\fB_\fIlast\fB_\^_\fIsuffix\fR\ .
See
.B first,
.B middle,
.B middle2,
.B last,
and
.B suffix
for the characteristics of the individual parts of
.BR pn " and
.BR name .
.PD
.SS Characteristics
.PD 0
.TP
.B a
Ignore non-alphanumerics.
.TP
.B i
Ignore case.
.TP
.B m
Multiple records match.
If argument is empty,
.BR 1 " or
.BR 2 ,
retrieve first, second or third record for
.IR query ,
respectively.
.PD
.PP
\*p attributes used for mail delivery are described in the following table.
These attributes only take effect on
.IR queries .
.PP
.TS
center,box,tab(;); 
c s s
c | c | c 
c | l | lw(3i).
\fBDelivery Attributes\fP
_
\fIAttribute;Possible Values;Description
=
\fRdel;paper;T{
Delivery type.
Forces paper mail delivery for \fIquery\fP recipients.
T}
_
ask;yes, no (default);T{
Ask before sending mail.
The recipients of the message are displayed.
T}
_
all;yes, no (default);T{
Send to all recipients matching \fIquery\fP.
Useful when sending to entire organizations.
The \fIquery\fP is not expanded in the message header.
T}
_
.TE
.SS Output Formatting
The
.I fmt
argument specified with the
.B \-o
option is used as a
.IR printf (3C)-like
string for formatting the output of directory queries.
All characters are copied literally, except for
attribute substitutions and backslash escapes.
Quoting may be necessary to prevent shell interpretation.
.PP
The syntax for attribute substitutions is:
.IP
\fB% \fIflags width . precision \fIattribute\fP
.PP
Only
.BR % " and
.I attribute
are required.
.I Flags
may be one or more of the following:
.RS 2
.PD 0
.TP
.B \(mi
Right justify (left justification is the default).
.TP
.B ^
Capitalize the first letter of each word.
.TP
.B +
Capitalize all letters in the value.
.TP
.B <
If the value is empty,
delete back to the last
.B \e\^<
or beginning of output.
.TP
.B >
If the value is empty, skip to the next
.B \e\^>
or end of format string.
.RE
.PD
.PP
.I Width
is an integer giving the minimum field width.  
If the value has fewer characters than this number, the
value will be padded on the right (or left, if the
.B \(mi
flag is present) with blank spaces.
The default
.I width
is zero. 
.P  
.I Precision
is an integer that specifies 
the maximum number of characters to
be printed from the value.
If the value has 
more characters than this number, they will not
be printed.
Zero is a special case meaning print all characters (the default).
A period is used to
separate 
.I width
and
.I precision
and is only used if
.I precision
is used.
.PP
To follow an
.I attribute
with an alphanumeric,
enclose the attribute substitution
(except for the percent sign) in curly braces,
.BR {} .
.PP
.ne 3i
The backslash escapes are:
.PP
.TS
allbox,center,tab(;);
c s
c | c
c | l.
\fBOutput Format Backslash Escapes\fP
\fIEscape;Meaning\fR
_
\e\^c;Suppress terminating new-line
\e\^b;Backspace
\e\^f;Form feed
\e\^n;New-line
\e\^r;Carriage return
\e\^t;Tab
\e\^v;Vertical Tab
\e\^\fInnn\fR;Octal ASCII character
\e\^<;Marker for < flag
\e\^>;Marker for > flag
.TE
.br
.SS Environment Variables
.TP
.B POSTETC
When this variable is set, \*p will search
.B /etc/passwd
for
.I login
arguments. 
.TP
.BI POSTQUAL\= query
Provide default
.I "value pairs"
for
.IR query .
If a
.I query
is missing
.I "value pair"
given in the
.B POSTQUAL
argument, add it to the
.I query
.RB ( \-q
overrides
.BR POSTQUAL ).
.SH EXAMPLES
.PD 0
Read from your \fBmbox\fP file:
.IP
post \-f
.PP
Send mail to all department heads in Division 452:
.IP
post org=452/tl=dh/all=yes
.PP
Send paper mail to all employees at Whippany in section 3A:
.IP
post loc=wh/room=3a/all=yes/del=paper
.PP
Query directory for employees using system whutt:
.IP
post \-w ema=whutt!*
.PP
Query directory and format output in label-like form:
.IP
post \-w \-o "%^pn\e\^n%+loc %+room" org=123
.PP
Query directory for valid attributes:
.IP
post \-w \-o %attribute
.PD
.SH FILES
.PD 0
.TP 30
$POST/lib/dispatch
distributed directory map
.TP
$POST/lib/mailx
mailx program used by \*p
.TP
$POST/lib/user
list of
.IR post (1)
users on this system
.TP
$POST/lib/dbinfo
program to retrieve directory information
.TP
$POST/lib/update
program to update directory
.TP
$POST/lib/genuser
program to update
.IR user (4)
file
.TP
/bin/mail
local mail delivery agent
.TP
/bin/rmail
remote mail delivery agent
.TP
/etc/passwd
list of local users
.PD
.SH SEE ALSO
btoa(1),
fixmbox(1),
forward(1),
gone.fishing(1),
mail(1),
mailx(1),
pq(1),
genesis(1M),
server(1M),
upas(1M),
pq(3C),
dispatch(4),
ev(4),
mail(4),
modules(4),
user(4).
.br
.nf
.I "Post 4.0 Product Release Description"
.I "Post 4.0 Primer"
.I "Post 4.0 User Guide"
.I "Post 4.0 User Reference Manual"
.I "Post 4.0 Administrator Reference Manual"
.I "Post 4.0 Quick Reference Card"
.I "Post 4.0 Programmer Guide"
.I "Post 4.0 Administrator Guide"
.fi
.SH WARNINGS
Colon must be used in place of slash in the construction:
\fIsystem\fB!\fIquery\fR.
.SH BUGS
Mailings to very large mailing lists, e.g.,
.BR com=bl/all=yes ,
may fail by exceeding system limits (see
.IR exec (2)).


V10/cmd/post/post.c
#include <stdio.h>
#include <string.h>
#include <malloc.h>
#include <pwd.h>

#define	OPTS	"a:m:o:q:wxyDSVdef:mpr#"
#define	MAILX	"/bin/mail"
#define	UPDATE	path("update")
#define	DBINFO	path("dbinfo")
#define	DELPAP	"del=p"
#define	ALLYES	"all=y"
#define	ASKYES	"ask=y"

#define	PREMA(e)	printf("%s%s", e, yflg ? ", " : "\n")
#define	PRMULT(m, p)	PREMA(m[*m[EMAIL] && !p ? EMAIL : PAPER])

#define	WHO	0
#define	EMAIL	1
#define	PAPER	2
#define	NARGS	3
#define	MULT	10

typedef char	*strings[512];
typedef char	buffer[BUFSIZ];

char	usage[] = "usage:\
	post [-a attr] [-m mods] [-o fmt] [-q query]\n\
	     [-w] [mail options] [addresses...]\n\
	     [-D] [-S] [-V]\n";

strings	qdef;
strings	postattr;
buffer	postofmt;
int	postetc;
int	wflg, xflg, yflg;
int	askyes = 0;
int	exitval = 0;
char	*mult[MULT][NARGS];
char	*pq;
void	resolve();

extern char	emp_fmt[], ema_fmt[], who_fmt[];
extern char	emp_def[], emp_seq[], emp_uni[];
extern char	*pq_open();
extern char	*getenv(), *strdup();
struct passwd	*getpwnam();

char	*ofmt = emp_fmt, *attr = emp_def, *mods = "", *qopt = "";

main(argc, argv)
int	argc;
char	*argv[];
{
	extern char	*optarg;
	extern int	optind;
	strings	args;
	buffer	buf;
	char	**ap = args;
	int	c;

	*ap++ = argv[0];

	while ((c = getopt(argc, argv, OPTS)) != EOF)
		switch (c) {
		case 'a':
			attr = optarg;
			break;
		case 'm':
			mods = optarg;
			break;
		case 'o':
			ofmt = optarg;
			break;
		case 'q':
			qopt = optarg;
			break;
		case 'w':
			wflg++;
			break;
		case 'x':
			xflg++;
			break;
		case 'y':
			yflg++;
			break;
		case 'D':
			execvp(DBINFO, argv);
			perror(DBINFO);
			return 1;
		case 'S':
			execvp(UPDATE, argv);
			perror(UPDATE);
			return 1;
		case 'V':
			puts("research post 4.0 9/1/88");
			return 1;
		case '?':
			fputs(usage, stderr);
			return 1;
		default:
			sprintf(*ap++ = malloc(3), "-%c", c);
			if (optarg)
				*ap++ = optarg;
		}

	if (!wflg && !xflg && !yflg) {
		while (optind < argc)
			*ap++ = argv[optind++];
		*ap++ = 0;
		execv(MAILX, args);
		perror(MAILX);
		return 1;
	}
	if (xflg || yflg) {
		strcpy(buf, ofmt);
		strcat(buf, ema_fmt);
		ofmt = buf;
	}

	postetc = !!getenv("POSTETC");
	strvec(attr, postattr, "/:");
	fmtcomp(postofmt, ofmt, qdef);
	smerge(getenv("POSTQUAL"), qdef);
	smerge(qopt, qdef);

	if (option(ASKYES, qdef))
		askyes++;
	if (xflg || yflg)
		smerge(emp_uni, qdef);

	if (wflg && optind == argc)
		argv[argc++] = "";
	while (optind < argc) {
		resolve(argv[optind++]);
		if (yflg)
			putchar('\n');
	}
	if (pq && pq_close(pq) == -1)
		pqerr("pq");
	return yflg ? 0 : exitval;
}

void
resolve(name)
char	*name;
{
	strings	q, n, rec;
	int	delpap, allyes;
	int	c, r;
	char	*p;
	buffer	out;

	p = strpbrk(name, "!@%/:=");
	if (*name && strchr("+|", *name) || p && !strchr("/:=", *p)
	 || !p && !strpbrk(name, "._") && postetc && getpwnam(name)) {
		PREMA(name);
		return;
	}

	if (!dodef(name, n, postattr)) {
		eomatch(0, 0, 0, name);
		return;
	}
	for (c = 0; q[c] = qdef[c]; c++);
	merge(n, q);
	delpap = option(DELPAP, q);
	allyes = option(ALLYES, q);
	if (option(ASKYES, q))
		askyes++;

	c = 0;
	if ((r = pqowrite(q)) != -1) {
		while ((r = pq_read(pq, rec)) > 0) {
			fmtexec(out, postofmt, rec);
			if (match(c++, out, delpap, allyes, name, q))
				return;
		}
		if (r == -1)
			pqerr(name);
	} else
		pqerr(name);

	if (c == 0 && r == 0)
		fprintf(stderr, "post: %s: Not found\n", name);
	eomatch(c, delpap, allyes, name);
}

dodef(name, vars, defs)
char	*name;
strings	vars, defs;
{
	static buffer	tmp, buf;
	register char	*bp = buf;
	register int	m, i, c;

	strcpy(tmp, name);
	strvec(tmp, vars, "/:");
	for (m = 0; defs[m]; m++);

	for (i = 0; vars[i]; i++)
		if (!strchr(vars[i], '='))
			if (i < m) {
				c = sprintf(bp, "%s=%s", defs[i], vars[i]);
				vars[i] = bp;
				bp += c + 1;
			} else {
				fprintf(stderr, "post: %s: %s: No attribute\n",
					name, vars[i]);
				return 0;
			}
	return 1;
}

match(c, out, delpap, allyes, name, q)
int	c;
char	*out;
int	delpap, allyes;
char	*name;
strings	q;
{
	buffer	buf;
	int	i, m = c % MULT;
	char	*s;

	if (wflg) {
		fputs(out, stdout);
		return 0;
	}

	s = strdup(out);
	for (i = 0; i < NARGS && (mult[m][i] = s); i++)
		if (s = strchr(s, '|'))
			*s++ = 0;
	skname(mult[m][PAPER]);
	if (s = strchr(mult[m][EMAIL], '('))
		if (xflg)
			*--s = 0;
		else
			prname(s);

	if (allyes) {
		if (xflg) {
			PRMULT(mult[m], delpap);
			return 0;
		}
		if (yflg) {
			s = buf;
			s += sprintf(s, "!");
			for (i = 0; q[i]; i++)
				if (strchr(q[i], '=') && strcmp(q[i], emp_uni))
					s += sprintf(s, "%s:", q[i]);
			if (delpap)
				s += sprintf(s, "%s:", DELPAP);
			strcpy(s, ALLYES);
			PREMA(buf);
			return 1;
		}
	}
	if (++m < MULT)
		return 0;
	return session(c, m, delpap, name);
}

eomatch(c, delpap, allyes, name)
int	c, delpap, allyes;
char	*name;
{
	if (wflg || (xflg && allyes && c > 0))
		return 1;
	if (c == 1) {
		PRMULT(mult[0], delpap);
		return 1;
	}
	session(c, c % MULT, delpap, name);
}

session(c, m, delpap, name)
int	c, m, delpap;
char	*name;
{
	strings	addr;
	buffer	ans;
	int	i;

	if (c > 0 && c < MULT)
		fprintf(stderr, "post: %s: Ambiguous\n", name);
	for (i = 0; i < m; i++)
		fprintf(stderr, "%d %s\n", i, mult[i][WHO]);
	if (xflg || !isatty(0))
		exit(1);

	for (;;) {
		fprintf(stderr, "Which one? [");
		if (m > 0)
			fprintf(stderr, "0-%d, ", m-1);
		if (m == MULT)
			fprintf(stderr, "CR(more), ");
		fprintf(stderr, "o(mit), new address(es), q(uit)] ");

		if (!fgets(ans, sizeof ans, stdin))
			exit(1);
		ans[strlen(ans)-1] = 0;
		if (strlen(ans) <= 1)
			switch (*ans) {
			case 'o':
				return 1;
			case 'q':
				exit(1);
			case 0:
				if (m == MULT)
					return 0;
				else
					continue;
			default:
				i = *ans - '0';
				if (i >= 0 && i < m) {
					PRMULT(mult[i], delpap);
					return 1;
				}
			}

		strvec(ans, addr, " \t\n");
		for (i = 0; addr[i]; i++)
			resolve(addr[i]);
		return 1;
	}
}

merge(qs, qd)
strings	qs, qd;
{
	int	i, j, l, m;
	char	*s;

	for (l = m = 0; qd[l]; l++, m++);
	for (i = 0; qs[i]; i++) {
		for (j = 0; j < l; j++)
			if ((s = strchr(qd[j], '='))
			 && !strncmp(qs[i], qd[j], s-qd[j]+1)) {
				qd[j--] = qd[--l];
				qd[l] = qd[--m];
			}
		qd[m++] = qs[i];
	}
	qd[m] = 0;
}

smerge(s, qd)
char	*s;
strings	qd;
{
	strings	qs;

	if (s) {
		strvec(s, qs, "/:");
		merge(qs, qd);
	}
}

option(a, q)
char	*a;
strings	q;
{
	char	*s = strchr(a, '=');
	int	i, m, rv = 0;

	for (m = 0; q[m]; m++);
	for (i = 0; i < m; i++)
		if (!strncmp(q[i], a, s-a+1)) {
			if (!strncmp(q[i], a, strlen(a)))
				rv++;
			q[i--] = q[--m];
		}
	q[m] = 0;
	return rv;
}

pqowrite(q)
strings	q;
{
	static	tried = 0;
	strings	argv;

	if (!tried++) {
		strvec(mods, argv, " \t\n");
		pq = pq_open(argv);
	}
	return pq ? pq_write(pq, q) : -1;
}

pqerr(msg)
char	*msg;
{
	extern char pq_error[];
	fprintf(stderr, "post: %s: %s\n", msg, pq_error);
	exitval++;
}

prname(s)
char	*s;
{
	for (; *s; s++)
		if (*s == '_') {
			if (s[1] == '_')
				*s++ = ',';
			*s = ' ';
		}
}

skname(s)
char	*s;
{
	char	*ns;

	for (ns = s; *s; s++)
		if (*s != ' ' && *s != '\'')
			*ns++ = *s;
	*ns = 0;
}
